Improve performance of the Logstash Pipeline Viewer

ycombinator opened the pull request:
_This PR targets a feature branch instead of `master`. Once this PR is merged, a follow up PR will be made to the feature branch to update tests. The idea here is to keep the size of each PR small. Finally, a PR will be created from the feature branch to `master`._

Resolves #27513.

## Before this PR
The Logstash Pipeline Viewer UI would make a single Kibana API call to fetch all the information necessary to render the Logstash pipeline. This included information necessary to render the detail drawer that opens up when a user clicks on an individual vertex in the pipeline.

Naturally, this single API call fetched _a lot_ of data, not just from the Kibana server but also, in turn, from Elasticsearch as well. The "pro" of this approach was that the user would see instantaneous results if they clicked on a vertex in a pipeline and opened the detail drawer for that vertex. The "cons" were the amount of computation Elasticsearch had to perform and the amount of data being transferred over the wire between Elasticsearch and the Kibana server as well as between the Kibana server and the browser.

## With this PR
This PR makes the Kibana API call to fetch data necessary for **initially** rendering the pipeline — that is, with the detail drawer closed — much lighter. When the user clicks on a vertex in a pipeline, a second API call is then made to fetch data necessary for the detail drawer.

Concretely, this PR makes the following changes:

* [x] For each metric of each vertex in the pipeline viewer, instead of returning a timeseries in the Kibana API response, we now only return the _latest_ data point. In other words, instead of returning `[ [t1, v1], [t2, v2], ... [tn, vn] ]` we now only return `vn`. This dramatically cuts down the number of aggregation buckets that Elasticsearch needs to generate.

* [x] For each vertex in the pipeline viewer, instead of returning up to 9 metrics in the Kibana API response we now only return 4. These are the 4 that are actually needed for the initial rendering of the pipeline (i.e. with the detail drawer closed). The 5 metrics that are no longer returned are `events_in`, `events_out`, `duration_in_millis`, `queue_push_duration_in_millis`, and `queue_push_duration_in_millis_per_event`.  

* [x] Since we no longer return `queue_push_duration_in_millis`, and `queue_push_duration_in_millis_per_event` from the Kibana API response, we also no longer compute the aggregations behind these metrics. Specifically, we no longer need the `queue_push_duration_in_millis_temp_by_ephemeral_id` and `queue_push_duration_in_millis_total` aggregations.

* [x] We no longer return the `pipeline.representation.plugins` array in the Kibana API response, since this array is not currently used in the Pipeline Viewer UI. We do this by excluding the `plugins` field from the Elasticsearch API response for the `logstash_state` document.

* [x] A new API to return timeseries metrics for a single vertex is created. This API is called whenever the detail drawer is open and showing a plugin vertex.

## Gains, by the numbers

Based on a simple, 1-input, 1-filter, and 1-output pipeline.

* Before this PR, the Elasticsearch `logstash_stats` API responses (multiple calls were made using the `composite` aggregation over the `date_histogram` aggregation) generated a total of 1228 aggregation buckets (before any `filter_path`s were applied but across all `composite` "pages"). With this PR, the single `logstash_stats` API response (note that this is just for the initial rendering of the pipeline, with the detail drawer closed) generated 12 buckets (also before any `filter_path`s were applied). That's a **99.02% reduction** in number of buckets.

* Before this PR, the Elasticsearch `logstash_stats` API responses added up to 70319 bytes. With this PR, the single `logstash_stats` API response for the same pipeline is 746 bytes. That's a **98.93% reduction** in size.

* Before this PR, the Elasticsearch `logstash_state` API response was 7718 bytes. With this PR, the API response for the same pipeline is 2328 bytes. That's a **69.83% reduction** in size.

* Before this PR the Kibana API response was 51777 bytes. With this PR, the API response for the same pipeline is 2567 bytes (again, note that this is just for the initial rendering of the pipeline, with the detail drawer closed). That's a **95.04% reduction** in size.

## Testing this PR

1. [Setup Logstash Monitoring](https://www.elastic.co/guide/en/logstash/current/configuring-logstash.html).
2. Start a Logstash pipeline, like so:

   ```
   ./bin/logstash -e 'input { generator {} } filter { drop {} } output { stdout { codec => dots } }'
   ```
3. Let the pipeline run for ~30 minutes (due to [this bug](https://github.com/elastic/kibana/issues/28153) 😞).

4. Open the Monitoring UI and visit the Logstash Pipeline Viewer for your pipeline.

5. Make sure the pipeline shows up. Check that the API response for the initial rendering of the pipeline doesn't contain any timeseries data. Then click on a vertex to open the Detail Drawer. Check that the details for that vertex show up, including sparklines rendered using timeseries data. Sample screenshots of what to expect:

   * Initial rendering

   <img width="1182" alt="Screen Shot 2019-03-18 at 11 02 45 AM" src="https://user-images.githubusercontent.com/51061/54552538-87caf980-496d-11e9-93ad-ede1eb6defa9.png">

   * With detail drawer open for `drop` filter plugin vertex

   <img width="1201" alt="Screen Shot 2019-03-18 at 11 02 57 AM" src="https://user-images.githubusercontent.com/51061/54552548-8dc0da80-496d-11e9-9e9e-8d130ed361ca.png">

ycombinator committed:
WIP: Trying out a couple things

1. Using a time series interval of 10s is pretty meaningless because it's also the collection interval. So each bucket ends up being 1 doc, and any difference-based stats end up being 0. So trying to increase the min. time series interval to 30s. Obviously not something we want to hardcode but an interesting experiment.

2. To reduce the number of buckets being computed by ES (and therefore returned over the wires), what if we set a range filter on the timestamp field such that we only look at the latest couple of time buckets' worth of docs?

ycombinator committed:
[WIP] Almost done with server-side changes

ycombinator committed:
[WIP] Getting the config view working

elasticmachine commented:
Pinging @elastic/stack-monitoring

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/931/)

ycombinator committed:
Don't compute aggs for queue push duration

ycombinator committed:
Only return necessary metrics from Kibana API response

ycombinator committed:
Don't bother fetching plugins from ES doc as they aren't yet being used

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/1069/)

Dumbaz commented:
Hey, nice to see Progress on the Monitoring Features.

Will there possibly be a backport to the 6.X branch of Kibana? As with most features in the Monitoring Feature, we need a fix sooner than later.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4522/)

ycombinator committed:
WIP: Getting single vertex stats agg working

ycombinator committed:
WIP: Get single vertex stats API to return desired structure

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/4531/)

ycombinator committed:
Merge APIs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5224/)

ycombinator committed:
Refactoring to pull up detail vertext state

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5514/)

ycombinator committed:
WIP: Get detail drawer to open/close right away on clicks

ycombinator committed:
Make vertex model from API response

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5627/)

ycombinator commented:
@chrisronline Quick question about this PR. I'm getting pretty close to wrapping up the changes made by this PR and hope to have them done today. At that point, what will be left is to update the tests.

Given the size of this PR, are you okay with me adding the test updates to this PR? Alternatively, I could re-target this PR to a feature branch and then make a follow up PR to the same branch with just test updates, and then finally make a PR from the feature branch to master for everything (which shouldn't need thorough review, just green CI).

chrisronline commented:
@ycombinator Let's do the feature branch approach

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5827/)

ycombinator committed:
WIP: Making eps sparkline work in detail drawer

ycombinator committed:
Remove logic that should be more common

ycombinator committed:
Clear out detailVertexId if pipeline version is changed

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5864/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5901/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/5906/)

chrisronline reviewed (commented):
Code, at a high level, looks good! I'm going to make a pass on functionality now

ycombinator commented on x-pack/plugins/monitoring/server/lib/logstash/get_pipeline_vertex.js:
```diff
@@ -0,0 +1,123 @@
+  // const totalDurationStats = vertexStatsAggregation.aggregations.pipelines.scoped.total_processor_duration_stats;
+  // const totalProcessorsDurationInMillis = totalDurationStats.max - totalDurationStats.min;
+  // TODO
+  const totalProcessorsDurationInMillis = 100000;
```
This TODO needs to be fixed.

ycombinator commented on x-pack/plugins/monitoring/server/lib/logstash/get_pipeline_vertex_stats_aggregation.js:
```diff
@@ -0,0 +1,182 @@
+  return {};
+  return {
+    total_processor_duration_stats: {
+      // TODO
```
This TODO needs to be fixed (related to the other TODO).

chrisronline reviewed (approved):
This looks great! I can't really attest to the benefit at scale, but everything seems to work fine locally for me. I made one small note fyi.

ycombinator commented on x-pack/plugins/monitoring/public/views/logstash/pipeline/index.js:
```diff
@@ -28,6 +29,9 @@ import {
 } from '@elastic/eui';
 
+let previousPipelineHash = undefined;
+let detailVertexId = undefined;
```
Yes, this was not my intention and I agree that clearing out `detailVertexId` is safer. Will fix, thanks for catching!

justinkambic commented:
I did a first pass at a functional review earlier today with the basic pipeline recommended in the description.

- > Check that the API response for the initial rendering of the pipeline doesn't contain any timeseries data.

  Looks good.

  ![image](https://user-images.githubusercontent.com/18429259/54622868-cb426800-4a40-11e9-8188-f6a365ae7ced.png)

- > Then click on a vertex to open the Detail Drawer. Check that the details for that vertex show up, including sparklines rendered using timeseries data.

  Also looking good!

  ![image](https://user-images.githubusercontent.com/18429259/54623117-4dcb2780-4a41-11e9-923c-5718161999fa.png)

ycombinator committed:
Implementing TODOs

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6069/)

ycombinator committed:
Use this.updateData

ycombinator committed:
Reset variables when leaving page

ycombinator commented:
@chrisronline I know you already gave your LGTM but that was before I'd addressed your review feedback. I've addressed it now, in case you want to take another look.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/6090/)

chrisronline commented:
@ycombinator These test failures are directly related to this change right?

> The 5 metrics that are no longer returned are events_in, events_out, duration_in_millis, queue_push_duration_in_millis, and queue_push_duration_in_millis_per_event.

ycombinator commented:
@chrisronline Yes, looking at the failures in CI, that does appear to be the case.

justinkambic commented:
@ycombinator what's the desired behavior when a cluster isn't found? If the URL has a non-existent UUID, the server will log:
```
server   error  [13:43:23.168] [error][monitoring] TypeError: Cannot read property 'firstSeen' of undefined
    at getPipeline (/Users/jk/git/justinkambic/kibana/x-pack/plugins/monitoring/server/lib/logstash/get_pipeline.js:99:66)
    at handler (/Users/jk/git/justinkambic/kibana/x-pack/plugins/monitoring/server/routes/api/v1/logstash/pipeline.js:70:26)
    at process._tickCallback (internal/process/next_tick.js:68:7)
```
Would it be useful to add some handling to catch the case where the [version object](https://github.com/elastic/kibana/pull/31293/files#diff-6740aad2a5ce6f79c08b9dfd1a748813R99), or other related objects are undefined, and return a 404?

ycombinator commented:
@justinkambic 
> what's the desired behavior when a cluster isn't found? If the URL has a non-existent UUID, ...

1. This shouldn't happen in the normal course of user interactions with the UI. If a user manually changes the UUID in the URL to something non-existent and we fail in an ugly way, I think that's okay because the user shouldn't be mucking around like that anyway. Curious: what steps did you take to end up in a state with a non-existent cluster UUID in the URL?

2. I just tested this behavior in `master` and we fail the same (ugly) way there as well. So this behavior isn't specific to this PR. If we want to address it, I'd open a new issue rather than increase the scope of this PR with a fix.

justinkambic commented:
> If we want to address it, I'd open a new issue rather than increase the scope of this PR with a fix.

I agree if we wanted to fix it it should happen in a subsequent change.

> If a user manually changes the UUID in the URL to something non-existent and we fail in an ugly way, I think that's okay

👍 your assessment is fine by me, thanks!

justinkambic reviewed (approved):
LGTM!

chrisronline reviewed (approved):
LGTM, once CI is green! Nice work!

ycombinator commented:
> LGTM, once CI is green! 

CI won't go green on this PR as tests will fail. Per our [earlier discussion](https://github.com/elastic/kibana/pull/31293#issuecomment-473952483), we agreed to defer tests to a [separate PR](https://github.com/elastic/kibana/pull/33570).

So, two options:

1. if you want CI to go green on this PR, we'd need to pull in the tests from the other PR into this one, OR 
2. we stick to our original plan of merging this PR (into a feature branch) with CI failing, knowing that a follow up PR will make tests pass and CI green.

Can you clarify which way you want to proceed?

chrisronline commented:
I misunderstood. Let’s continue with option 2 as planned.
